examples/gno.land/r/gov/dao/proxy/types.gno

```diff
@@ -0,0 +1,131 @@
+package proxy
+
+import (
+	"errors"
+	"std"
+)
+
+type ProposalID int
+
+// VoteOption is the limited voting option for a DAO proposal
+// New govDAOs can create their own VoteOptions if needed in the
+// future.
+type VoteOption string
+
+const (
+	YesVote     VoteOption = "YES"     // Proposal should be accepted
+	NoVote      VoteOption = "NO"      // Proposal should be rejected
+	AbstainVote VoteOption = "ABSTAIN" // Side is not chosen
+)
+
+type VoteRequest struct {
+	Option     VoteOption
+	ProposalID ProposalID
+	Metadata   Metadata
+}
+
+type ProposalRequest struct {
+	Title       string
+	Description string
+	Callback    func(Metadata) error
+	Metadata    Metadata
+}
+
+type Proposal struct {
+	Author Author
+
+	Title       string
+	Description string
+
+	Executor    *executor
+	AllowedDAOs []string
+}
+
+// Author is an interface returned by the current govDAO implementation
+// representing the creator of a proposal. It is an interface
+// because we don't really know who/what will be allowed to create proposals
+// in the future, they can be any entity.
+type Author interface {
+	ID() string
+	String() string
+}
+
+type executor struct {
+	metadata   Metadata
+	callbackFn func(m Metadata) error
+}
+
+func newExecutor(fn func(m Metadata) error, metadata Metadata) *executor {
+	return &executor{
+		metadata:   metadata,
+		callbackFn: securityWrapper(fn),
+	}
+}
+
+func (e *executor) String() string {
+	if e.metadata == nil {
+		return "No metadata available."
+	}
+
+	return e.metadata.String()
+}
+
+func (e *executor) Execute() error {
+	return e.callbackFn(e.metadata)
+}
+
+func securityWrapper(fn func(m Metadata) error) func(m Metadata) error {
+	return func(m Metadata) error {
+		if index(std.PreviousRealm().PkgPath()) < 0 {
+			return ErrInvalidExecutorCaller
+		}
+
+		return fn(m)
+	}
+}
+
+var ErrInvalidExecutorCaller error = errors.New("the realm calling the proposal executor was not a govDAO realm")
+
+// Metadata contains any kind of information that will be stored on the executor
+// and will be sent to the callback function when the proposal is validated.
+// It can contains any kind of data.
+type Metadata interface {
+	IsMetadata()
+	String() string
+}
```

Comment:
Also, an opinionated alternative could be something in a "similar" fashion as the error interface:

```go
type Metadata interface {
  Metadata() string
}
```

If a smaller interface would make sense.